### PR TITLE
Add batch normalization layer and mean/sum tensor operations

### PR DESCRIPTION
**Pull request by workofart**
## Description
- Add batch norm layer (forward implementation, backward implementation covered by Tensor primitive backward operations)
- Add mean/sum tensor operations to support batch norm operations

## Tests

Added unit tests for batch norm, mean, sum.

Updated mnist.py to compare with or without batch norm and Adam vs SGD.
image

image

### CODE REVIEW

File: autograd/tensor.py [hunk 1]
```diff
@@ -17,7 +20,13 @@ def __init__(self, data, prev=(), requires_grad=True):
     # For each of these primitive operations, we need to adjust the backward gradient computation accordingly
     def __add__(self, other):
         if not isinstance(other, Tensor):
-            other = Tensor(other)
+            # Ensure other is converted to a Tensor with the same shape as self
+            other = Tensor(
```
Comment (workofart) [inline autograd/tensor.py:24]: This is to handle the self + scalar addition case more gracefully.

File: autograd/tensor.py [hunk 2]
```diff
@@ -191,9 +200,24 @@ def _backward():
...
             self.grad += other.data * (self.data ** (other.data - 1)) * result.grad
-            if self.data > 0:
-                other.grad += (self.data**other.data) * np.log(self.data) * result.grad
+
+            # Gradient w.r.t exponent (other)
+            valid_base = self.data > 0
+            grad_y = (self.data**other.data) * np.log(np.abs(self.data)) * result.grad
+
+            # Handle scalar and array cases
+            if np.isscalar(other.data):
```
Comment (workofart) [inline autograd/tensor.py:211]: Similarly, handle scalar and array cases for the `__pow__` primitive operation.

File: autograd/tensor.py [hunk 3]
```diff
@@ -210,6 +234,105 @@ def __neg__(self):
...
+            prev=(self,),
+            requires_grad=self.requires_grad,
+        )
+
+        result._backward = lambda: self._reduce_ops_backward(
+            output=result, axis=axis, keepdims=keepdims
+        )
+        return result
+
+    def _reduce_ops_backward(self, output, axis=None, keepdims=False):
```
Comment (workofart) [inline autograd/tensor.py:297]: Creating a general backward function for reduce ops, since they typically share the same backward behavior with the element count used to scale the gradient value, and special handling for `axis` and `keepdims` cases.

File: examples/mnist.py [hunk 4]
```diff
@@ -1,5 +1,5 @@
 from autograd import nn, optim, functional, utils
-from sklearn.datasets import fetch_openml
+from openml.datasets import get_dataset
```
Comment (workofart) [inline examples/mnist.py:2]: switch to use openml library directly, which has better caching behavior
